ESPS/general/src/splay/splay.1
.\" Copyright (c) 1992 Entropic Research Laboratory, Inc.; All rights reserved
.\" @(#)splay.1	1.4 3/28/97 ERL
.ds ]W (c) 1992 Entropic Research Laboratory, Inc.
.TH SPLAY 1\-ESPS 9/11/91
.SH NAME
splay - send ESPS sampled data to the SPARCstation codec output.
.SH SYNOPSIS
.B splay
[
.BI \-x " debug-level"
] [
.BI \-{ps} " range"
] [
.BI \-g " gain"
] 
[
.BI \-R " nrepeat"
] [
.BI \-b " shift-value"
] [
.B \-a
] [
.B \-e
] [
.B \-q
]
.I file
[
.I more-files
]
.SH DESCRIPTION
.PP
.I Splay
does a mu-law encoding of the data in one or more ESPS, SIGnal, NIST
or headerless type SHORT sampled data
files and sends the data to the SPARCstation 1 digital-to-analog
converter.  A subrange of data within the files may be chosen; this
subrange may be specified in seconds or points. Furthermore, a repeat
count, a scaling factor (gain), and a shift factor may be specified.
.PP
The mu-encoding algorithm expects data in the range +/- 8059.  The
scaling, shifting, or automatic (see \fB\-a\fP) option can be used to
change the data values before mu encoding. If after all shifting or
scaling of the data is completed, there are still data values greater
than 8059, these values are set to 8059 and mu encoding is carried
out.
.PP
The SPARCstation supports only one sampling rate: 8000
samples/second.  \fISplay\fP checks the sampling rate of the input
file.  If it is different than 8000, it sends a warning to standard
error, resamples the data using a relatively crude but fast algorithm,
and plays the resampled data.  
.PP
Playback may be aborted by 
sending the terminal's interrupt character (normally control-C)
after playback has started. 
.PP
If "-" is given for a filename, then the input is taken from standard
input and must be an ESPS FEA_SD file (ie., SIGnal or NIST/Sphere
files cannot be used with standard input).  

.SH ENVIRONMENT
.PP
\fBSPLAY_GAIN\fP - audio chip playback volume level: from 0 to 100.
.br
.PP
\fBSPLAY_X\fP - if set to a non-zero value, use external speaker jack. If not set, use internal speaker. This is overrriden by the \fB\-e\fP option.
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-x " debug"
Only debug level 1 is defined in this version.   This causes several
messages to be printed as internal processing proceeds.   The default
is level 0, which causes no debug output.
.TP
.BI \-p " range"
Select a subrange of points to be played, using the format
.I start\-end
,
.I start:end 
or
.IR start:+count .
Either the start or the end may be omitted; the beginning or the end of the
file are used if no alternate is specified. 
.IP
If multiple files were specified, the same range from each file is played.
.TP
.BI \-r " range"
\fB\-r\fP is a synonym for \fB\-p\fP.
.TP
.BI \-s " range"
Select a subrange of speech, specified in seconds. 
The beginning of the file is 0.  The format for a subrange
is the same as the \fB\-p\fR option.  The arguments to the \fB-s\fR
option can be decimal numbers (e.g. 1.5).
.TP
.BI \-R " count"
Repeat count. The selected speech is played
.I count
times.  Playback from standard 
input can not be repeated.
If multiple files were specified, the files are played in order, and
the order is repeated
.I count
times.
.TP
.BI \-g " gain"
Speech samples are multiplied by
.I gain
before being mu-law encoded and played back.  Gain is a floating point value.
.TP
.BI \-b " shift-value"
If this option is given and \fIshift-value\fR is positive, then each
sample (after it is converted to integer format) is shifted to the right
by \fIshift-value\fR bits.   \fIShift-value\fR being negative causes a
left shift.   
The shifting takes place before the mu-law conversion.
This
is preferred to using the \fB-g\fR option because
the shift is faster than a floating multiply.   This difference is most
noticeable on large files.
.TP
.B \-a
If set and the generic header item \fImax_value\fP is set in the input
file header, automatic scaling is done.
This multiplies each data value by 8059/\fImax_value\fP before
mu encoding.
.TP
.B \-e
If set, the external output jack is used.
.TP
.B \-q
Suppresses various warnings and information messages (quiet mode). 
.PP
.SH "INTERACTION WITH XWAVES"
.I Splay
is designed to optionally use the server mode of \fIxwaves\fP(1\-\s-1ESPS\s+1).
This
is especially handy when
.I splay
is used as an \fIxwaves\fP external play
command (e.g. by setting the \fIxwaves\fP global play_prog).  When the latter 
is the case,
play commands initiated via \fIxwaves\fP' menu operations may be interrupted
by pressing the left mouse button in the data view.  \fIXwaves\fP will send
a signal 30 (SIGUSR1) to the play program.  Splay responds to this
by sending back to \fIxwaves\fP a command "set da_location xx", where xx is
the sample that was being output when play was interrupted.  This
setting, in conjunction with \fIxwaves\fP' built-in callback procedure for
handling child-process exits, causes the \fIxwaves\fP signal display to
center itself on the sample where play was halted.
.PP
The SIGUSR1 signal to terminate
.I splay
may come from any source.  If
it comes from sources other than \fIxwaves\fP, the environment variables
WAVES_PORT and WAVES_HOST must be correctly defined (see
\fIespsenv\fP(1-ESPS)), for correct functioning of the \fIxwaves\fP view
positioning.  (Of course, \fIxwaves\fP must actually be displaying the
signal in question at the time and \fIxwaves\fP must have initiated the
play.)
.PP
.I Splay
may also be interrupted with kill -2 (SIGINT) or kill -3
(SIGQUIT).  These signals are caught gracefully and
.I splay
halts immediately, but no message is sent to \fIxwaves\fP.  No message is sent
if the play operation finishes without interruption.
.PP
.SH ESPS PARAMETERS
.PP
The parameter file is not read.  
.SH ESPS COMMON
.PP
ESPS Common processing may be disabled by setting the environment variable
USE_ESPS_COMMON to "off".  The default ESPS Common file is .espscom 
in the user's home directory.  This may be overridden by setting
the environment variable ESPSCOM to the desired path.  User feedback of
Common processing is determined by the environment variable ESPS_VERBOSE,
with 0 causing no feedback and increasing levels causing increasingly
detailed feedback.  If ESPS_VERBOSE is not defined, a default value of 3 is
assumed.
.PP
If Common file processing is enabled 
the following are read from the Common file.
.IP
.I "filename - string"
.IP
This is the name of the input file.  If no input file is 
specified on the command line, \fIfilename\fP is taken to be the
input file.  If an input file is specified on the command line, 
that input file name must match \fIfilename\fP or the other 
items (below) are not read from Common. 
.IP
.I "start - integer"
.IP
This is the starting point in the input file to begin output.  It 
is not read if the \fB\-p\fP or \fB-s\fR option is used.  
.sp
.I "nan - integer"
.IP
This is the number of points to output from the input file.  It is not
read if the \fB\-p\fP or \fB-s\fR option is used.  A value of zero means the last
point in the file.
.sp
.PP
Again, the values of \fIstart\fR and \fInan\fR are only used if the
input file on the command line is the same as \fIfilename\fP in the
common file, or if no input file was given on the command line.  If
\fIstart\fR and/or \fInan\fR are not given in the common file, or if
the common file can't be opened for reading, then \fIstart\fR
defaults to the beginning of the file and \fInan\fR defaults to the
number of points in the file.
.PP
The following items are written into the ESPS Common file:
.IP
.I "start - integer"
.IP
The starting point from the input file.
.sp
.I "nan - integer"
.IP
The number of points in the selected range.
.sp
.I "prog - string"
.IP
This is the name of the program (\fIsplay\fP in this case).
.sp
.I "filename - string"
.IP
The name of the last input file.  
.PP
These items are not written to ESPS COMMON if the input file is <stdin>.
.SH ESPS HEADER
.PP
The generic header items \fImax_value\fP and \fIrecord_freq\fP are
read, if they exist. If the \fB\-a\fP is specified and \fImax_value\fP
exists, the value of the \fImax_value\fP header item is used
to automatically scale the data for full scale output. The \fIrecord_freq\fP
header item is checked to see if it equals 8000, if it doesn't,
a warning is issued.
.SH DIAGNOSTICS
.PP
.I Splay
informs the user if the input file does not exist, it is not an ESPS sampled
data file, the sampling frequency of the input file(s) != 8000, or if inconsistent options are used.
.PP
If the starting point requested is greater than the last point in the
file, then a message is printed and the program exits with status 1.  If
the ending point requested is greater than the last point in the file,
it is reset to the last point, a warning is printed and processing
continues.
.SH SEE ALSO
.nf
\fIesps2mu\fP(1\-\s-1ESPS\s+1), \fItofeasd\fP(1\-\s-1ESPS\s+1)
.fi
.SH AUTHOR
.PP
Program by Derek Lin.  Manual page by David Burton, modified by Derek Lin.
The program used a resampling funcion from Tom Veatch, Department
of Linguistics, University of Pennsylvania.  Thanks to Tom for
permission to use the resampling function. 


ESPS/general/src/splay/splay.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1990-1992 Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  Derek Lin
 * Checked by:
 * Revised by:
 *
 * Brief description:
 *
 */

static char *sccs_id = "%W%	%G%	ERL";


#include <stdio.h>
#include <math.h>
#include <fcntl.h>
#include <signal.h>
#include <esps/esps.h>
#include <esps/epaths.h>
#include <esps/fea.h>
#include <esps/feasd.h>
#include <esps/ss.h>
#include <Objects.h>

#define SYNTAX USAGE("splay [-x debug-level] [-{ps} range] [-g gain] [-R count] [-b shift-value] [-a] [-e] file [more-files]")

#define SPARC_SAMPLE_FREQ 8000
#define REALLY_HUGE_INTEGER 1000000000
#define DEBUG 0

int debug_level = 0, max_buff_bytes = 2000000, w_verbose = 0;
int da_location = 0;                    /* Waves variable */
int gflag = 0, bflag =0, aflag =0;      /* gain, shift, auto-scaling flag */
float gain;                             /* command-line gain*/
unsigned int gain_codec = 0;            /* codec gain */
int gain_codecflag = 0;
int left= 0; right = 0;                 /* left and right shifts */
int external = 0;                       /* external jack */
short get_fea_type();

main(ac,av )
     int ac;
     char **av;
{
  extern int optind;
  extern char *optarg;

  int ifd = 0;                          /* input file descriptor */
  char *ifile = NULL;                    /* input file name */
  int doseek = 0;                       /* samples to skip for istr */
  int sizeof_sample = sizeof(short);    /* sample size */
  int amp;                              /* signal amplitude */
  int length;                           /* file length */
  int noheader = FALSE;                 /* esps header signal */
  double srate = 8000;                   /* sampling rate */
  double sstart, start = -1.0, end = -1.0, tmp;  /* start/end time */
  long start_p = 1, end_p;               /* start/end sample point */
  char *r_switch = NULL, *s_switch = NULL; /* range switch */
  int fnum;
  int repeats = 1;
  int bits;
  int c;
  int common_range = 0;
  int espsfile = 0;
  int qflag = 0;
  int type;

#if !DEBUG
  while((c = getopt(ac,av,"p:s:r:R:g:x:b:qae?")) != EOF) 
    {
      switch(c) 
	{
	case 'x':
	  debug_level = atoi(optarg);
	  break;
	case 'g':
	  gain = atof(optarg);
	  gflag++;
	  break;
	case 'R':
	  repeats = atoi(optarg);
	  break;
	case 'r':
	case 'p':
	  if(s_switch){
	    fprintf(stderr,"Cannot use -r/-p with -s\n"); exit(1);}
	  r_switch = optarg;
	  break;
	case 's':
	  if(r_switch){
	    fprintf(stderr,"Cannot use -s with -r/-p\n"); exit(1);}
	  s_switch = optarg;
	  break;
	case 'e':
	  external = 1;
	  break;
	case 'a':
	  aflag++;
	  break;
	case 'b':
	  bits = atoi(optarg);
	  bflag++;
	  break;
	case 'q':
	  qflag++;
	  break;
	case '?':
	default:
	  SYNTAX;
	  break;
	}
    }
#endif

#if DEBUG
  debug_level = 3;
  fprintf(stderr,"debug_level %d\n",debug_level);
  ac = 1;
  optind = 1;
  external = 1;
  bflag = 1;
  bits = 3;
#endif

#if !DEBUG
  if( optind >= ac )
    {
      (void) read_params((char*)NULL, SC_CHECK_FILE, (char*)NULL);
      if(symtype("filename") == ST_UNDEF){
	fprintf(stderr,"%s: no input file\n", av[0]);
	SYNTAX;      /* exit */
      }
      ifile = getsym_s("filename");
      if(symtype("start") != ST_UNDEF || symtype("nan")!= ST_UNDEF)
	common_range = 1;
    }
  if(!external && getenv("SPLAY_X")) external = atoi(getenv("SPLAY_X"));
  if(getenv("SPLAY_GAIN")) {
    gain_codecflag++;
    gain_codec = atoi(getenv("SPLAY_GAIN"));
  }
#endif

  do {
    fnum = optind;
    do {				/* for all input specifications... */
      Header *h;                         /* SIG file header */
      Signal *sig;                       /* SIG file type */
      FILE *istr = stdin;                /* file stream */
      int hseek = 0;                     /* header bytes to skip */
      struct header *ehead = NULL;       /* esps header */
      
      if(ac > fnum)
	ifile = av[fnum];

      if(ifile && !strcmp(ifile,"-"))
	{
	  if(repeats > 1){
	    fprintf(stderr,"Can't replay for stdin file\n");
	    exit(1);
	  }
	  ifile = NULL;
	  ifd = 0;
	  istr = stdin;
	}
      if(ifile && *ifile)
	if((ifd = open(ifile, O_RDONLY)) < 0) 
	  {
	    fprintf(stderr,"Can't open %s for input.\n",ifile);
	    exit(-1);
	  }
      if(ifd > 0)			/* i.e. input is not stdin */
	if(!(istr = fdopen(ifd,"r"))) 
	  {
	    fprintf(stderr,"Can't open a stream from %d\n",ifd);
	    exit(-1);
	  }
      
      if(!noheader && (h = (Header*)get_header(ifd)) &&
	 (sig = (Signal*)new_signal(ifile,ifd,h,NULL,0,0.0,0))) 
	{
	  /* Input is ESPS file */
	  if(h->magic == ESPS_MAGIC)
	    {
	      espsfile = 1;
	      *(sig->smax) = *(sig->smin) = get_genhd_val("max_value",h->esps_hdr,0.0);
	      istr = h->strm;
	      hseek = h->esps_nbytes;
	      if (h->esps_hdr->common.type == FT_SD) 
		ehead = (struct header*)sdtofea(h->esps_hdr);
	      else
		ehead = h->esps_hdr;
	      
 	      switch(get_fea_type("samples", ehead)) {
	      case BYTE:
		sizeof_sample = sizeof(char);
		break;
	      case SHORT:
		sizeof_sample = sizeof(short);
		break;
	      case CHAR:
		sizeof_sample = sizeof(char);
		break;
	      case FLOAT:
		sizeof_sample = sizeof(float);
		break;
	      case DOUBLE:
		sizeof_sample = sizeof(double);
		break;
	      default:
		fprintf(stderr,"Unknown data type -- exiting\n");
		exit(1);
	      }
	      srate = *(double *) get_genhd_d("record_freq",ehead);
	      if( 1 != get_fea_siz("samples", ehead, (short *) NULL,
				   (long **) NULL)){
		fprintf(stderr,"Multichannel data not supported.\n");
		exit(1);
	      }
	      if( is_field_complex(ehead,"samples") == YES){
		fprintf(stderr,"Complex data not supported\n");
		exit(1);
	      }
	    }
	  else 
	    {		/* might be a SIGnal file... */
	      if((type_of_signal(sig) & VECTOR_SIGNALS) != P_SHORTS)
		{
		  fprintf(stderr,"Can't play signals of type 0x%x %d\n",
			  type_of_signal(sig));
		  exit(-1);
		}
	      hseek = sig->bytes_skip;
	      srate = sig->freq;
	      sizeof_sample = sizeof(short);
	    }

	  /* Position file stream to data */
	  if(ifd)
	    {
	      if(debug_level)
		fprintf(stderr,"Seeking in header setup\n");
	      fseek(istr, hseek, 0);
	    }
	  
	  if((amp = ((*sig->smax > -(*sig->smin))? *sig->smax : - *sig->smin))
	     < 128) amp = 32767;
	  
	  if((length = sig->file_size) <= 0) 
	    {
	      if(debug_level)
		fprintf(stderr,"No length info available in header.\n");
	      length = REALLY_HUGE_INTEGER;
	    }
	  
	  if (bflag)
	    if (bits < 0) left = bits*-1;
	    else right = bits;
	  
	  if (r_switch) 
	    {
	      end_p = length;
	      lrange_switch (r_switch, &start_p, &end_p, 1);
	      if (end_p > length){
		end_p = length;
		if(!qflag){
		  fprintf(stderr,"%s: end point (%d) > signal length\n",
			  av[0], end_p);
		  fprintf(stderr,"%s: end point reset to last point.\n");
		}
	      }
	      if(!end_p && length) 
		{ /* end not specified; play to end of file */
		end_p = length;
		length -= start_p;
	      }
	      length = end_p - start_p + 1;
	      if(debug_level)
		fprintf(stderr,"Range specified (-r); start_p:%d end_p:%d length:%d\n",
			start_p,end_p,length);
	      doseek = start_p-1;

	    } else if (s_switch){ 
	      frange_switch(s_switch, &start, &end);
	      if(((end < 0.0) || (end > start)) && (start > sig->start_time)) 
		{
		  doseek = 0.5 + (start - sig->start_time)*srate;
		  sstart = start;
		} else {
		  sstart = sig->start_time;
		  doseek = 0;
		}
	      length -= doseek;
	      if((end > sig->start_time) && (end > start) &&
		 (end < (tmp = (sig->start_time +
				((double)sig->file_size)/srate))))
		length -= ((tmp - end)*srate);
	    } else if(common_range){
	      if(symtype("start")!=ST_UNDEF)
		start_p = getsym_i("start");
	      if(symtype("nan")!=ST_UNDEF)
		end_p = start_p + getsym_i("nan");
	      if (end_p > length)
		end_p = length;
	      if(!end_p && length) 
		{ /* end not specified; play to end of file */
		end_p = length;
		length -= start_p;
	      }
	      length = end_p - start_p + 1;
	      if(debug_level)
		fprintf(stderr,"Common range specified: start_p:%d end_p:%d length:%d\n",
			start_p,end_p,length);
	      doseek = start_p-1;
	    }
	}
      else 
	{			/* Headerless file */
	  if(ifd) 
	    {	/* Is the input from a pipe? */
	      if(debug_level)
		fprintf(stderr,"Seeking after header setup\n");
	      fseek(istr,0,2);
	      if((length = ftell(istr)) > 10) {
		fseek(istr,0,0);
		length /= sizeof_sample;
		if((start>0.0)&&((start*srate)<length)) {
		  doseek = start*srate;
		  sstart = start;
		} else {
		  doseek = 0;
		  sstart = 0.0;
		}
		if((end > start) && ((end*srate) < length))
		  length = end*srate;
		length -= doseek;
		amp = 32767;
	      } else
		printf("Unreasonably small file (%s)\n",ifile);
	    } else {	      /* Can't actually seek or determine file size. */
	      if(start > 0.0) {
		char dummy[10000];
		int toskip = (0.5 + (start*srate))*sizeof_sample, skipped,
		da_location = (0.5 + (start*srate));
		
		while(toskip > 0) {
		  if((skipped = toskip) > 10000) skipped = 10000;
		  if((skipped = fread(dummy, 1, skipped, istr)) <= 0) {
		    fprintf(stderr,"Exhausted samples while seeking to requested start time (%f)\n",start);
		    exit(-1);
		  }
		  toskip -= skipped;
		}
	      }
	      length = REALLY_HUGE_INTEGER;
	      doseek = 0;
	    }
	  amp = 32767;
	}    /* end of handler for headerless input */
      
      if(doseek) {
	if(debug_level)
	  fprintf(stderr,"Seeking to sample # %d; sizeof_sample:%d\n",
		  doseek,sizeof_sample);
	da_location = doseek;
	fseek(istr, doseek*sizeof_sample, 1);
      }
      
      if(debug_level)
	fprintf(stderr,
		"Calling play_from_file. length=%d srate=%f amp=%d\n",
		length,srate,amp);
      if(length <= 0){
	fprintf(stderr,"The requested range is not present in the file, or the file has zero length.\n");
	exit(-1);
      }
      if( srate != SPARC_SAMPLE_FREQ && debug_level > 0 )
	fprintf(stderr,"%s: WARNING: sampling freq in %s != %d\n",
		av[0], ifile, SPARC_SAMPLE_FREQ);

      play_from_file(istr, length, &srate, amp, ehead,
		     (optind >= ac)? repeats : 1);
      fclose(istr);
    } while (ac > ++fnum);
  } while (--repeats && optind < ac);

  /* Write common file if it isn't standard input and it's esps file*/
  if(ifd !=0 && espsfile){
    (void) putsym_i("start", (int) start_p);
    (void) putsym_i("nan", (int) length);
    (void) putsym_s("filename", ifile);
    (void) putsym_s("prog", "splay");
    }
}


char    *ProgName = "splay";
char    *Version = "2.0";
char    *Date = "10/10/92";

void
set_pvd(hdr)
    struct header   *hdr;
{
    (void) strcpy(hdr->common.prog, ProgName);
    (void) strcpy(hdr->common.vers, Version);
    (void) strcpy(hdr->common.progdate, Date);

  }



